### PR TITLE
[PR] Phoenix 1.7rc upgrade

### LINKED ISSUE #54: Chore: Update Phoenix `1.7 Rc` ⬆️
With the release of **`Phoenix 1.7 Rc`** https://github.com/dwyl/learn-phoenix-framework/issues/152 we can now update this tutorial to include a few enhancements. 🪄 

# Todo

+ [ ] Go through the tutorial from _Scratch_ and re-create it with **`Phoenix 1.7 Rc`**

### PR DESCRIPTION
**Pull request by LuchoTurtle**
closes #54 

Still very much in progress. It's taking *much* longer than expected. The [migration guide](https://phoenixframework.org/blog/phoenix-1.7-released) isn't really helping much, as I'm constantly getting stuck with routing. 

The [`Routes.Helper`](https://www.solarwinter.net/unexpected-list/) isn't dynamically created no more and there are no instances of how to fix this in the guide, so I had to look in the docs for anything similar.

I'll keep on working 👍

### CODE REVIEW

**Comment by LuchoTurtle** [comment]
This is super frustrating.
I'm stuck on trying to render a component (`new.html.heex`). 

The old way was using `<%= render "form.html", Map.put(assigns, :action, Routes.item_path(@conn, :create)) %>
` but there I can't find **any** mention of rendering `.heex` files inside other `.heex` files (component reusability) 
https://hexdocs.pm/phoenix/1.7.0-rc.0/components.html

I really don't want to have to import `Phoenix.View` as dependency -> https://phoenixframework.org/blog/phoenix-1.7-released

I **shouldn't** need to... 😤

**Comment by LuchoTurtle** [comment]
Opened a question yesterday night but it was pending. 
Got some answers after it got accepted during the night so I might give these a whirl.

https://elixirforum.com/t/how-to-use-embed-heex-inside-heex-phoenix-1-7rc/52281

(although I think only one is viable)

**Comment by LuchoTurtle** [comment]
Got it to work. Explained it here -> https://elixirforum.com/t/how-to-use-embed-heex-inside-heex-phoenix-1-7rc/52281/4?u=luchoturtle.

Calling `<%= new(Map.put(assigns, :action, ~p"/items/new")) %>` works and embeds the form properly. 

Let's keep going...

**Comment by nelsonic** [comment]
Great that **`@t0t0`**'s answer on **`Elixir` Forum** was helpful. 🎉 
Really looking forward to seeing the progress on this PR. 👌 
It's _suuuuuuuper_ insightful to people considering using `Phoenix` how _much_ has changed over the past 18 months. 💭

**Comment by LuchoTurtle** [comment]
Calling actions from the view seems to be the bane of my existence. It's super demotivating when nothing seems to work.
I truly have no idea why they would remove the ability to use `Routes.item_path(@conn, :edit, item)` to call actions from the template. 
I *thought* `<%= delete(Map.put(assigns, :action, ~p"/items/#{item.id}")) %>`  would work but it is not and it puts into question how the other one worked. 
Spending *two* days on this is blasphemous and embarassing. Granted, the README will be updated accordingly but it's incredible how from 1.5 to 1.7 **nothing** works...

**Comment by LuchoTurtle** [comment]
...
And calling **actions** from the template.

Why does it work here?

```elixir
<.table id="items" rows={@items} row_click={&JS.navigate(~p"/items/#{&1}")}>
  <:action :let={item}>
    <.link href={~p"/items/#{item}"} method="delete" data-confirm="Are you sure?">
      Delete
    </.link>
  </:action>
</.table>
```

And not simply?
```elixir
   <.link
      class="destroy"
      navigate={~p"/items/#{item.id}"}
      method="delete"
    >
    </.link>
```

And if I want to use `<:actions>` by itself and try to make it work I get a 
`Invalid slot entry <:action>. A slot entry must be a direct child of a component`. Looking for this error leads me here https://github.com/phoenixframework/phoenix_live_view/issues/1892 but it really isn't all that helpful.

I mean, if I run `mix phx.routes`, I get the following output:
```
  GET     /                                      AppWeb.PageController :home
  GET     /items/toggle/:id                      AppWeb.ItemController :toggle
  GET     /items                                 AppWeb.ItemController :index
  GET     /items/:id/edit                        AppWeb.ItemController :edit
  GET     /items/new                             AppWeb.ItemController :new
  GET     /items/:id                             AppWeb.ItemController :show
  POST    /items                                 AppWeb.ItemController :create
  PATCH   /items/:id                             AppWeb.ItemController :update
  PUT     /items/:id                             AppWeb.ItemController :update
  DELETE  /items/:id                             AppWeb.ItemController :delete
```

Why isn't there a **simple way** of accessing `/items/:id` using the `delete` method? At this point I feel like I am almost forced to create a new route in `router` just to handle deleting files - but *this shouldn't happen*. It already works but why do I need to wrap anything in an element like a `<.form>` or `<.table>`? I shouldn't need to...

Sigh

**Comment by nelsonic** [comment]
Thank you for capturing your frustrations on this quest. 🙏
Keep going, you’re almost there! 🤞

**Comment by LuchoTurtle** [comment]
Finished updating all the README and all the permalinks in each step.
Haven't changed yet references to Heroku, though. 
Will get with deploying to Fly.io and update this readme accordingly. That way I should close two issues with this PR at once.

**Comment by LuchoTurtle** [comment]
Assigning his over to @nelsonic . Need to get this merged in order to deploy it to Fly.io

File: lib/app/mailer.ex [hunk 1]
```diff
@@ -0,0 +1,3 @@
+defmodule App.Mailer do
```
Comment (nelsonic) [inline lib/app/mailer.ex:1]: given that we aren't _using_ mailer in this app, could we just delete this file? ✂️

File: lib/app_web/components/core_components.ex [hunk 2]
```diff
@@ -0,0 +1,621 @@
+defmodule AppWeb.CoreComponents do
```
Comment (nelsonic) [inline lib/app_web/components/core_components.ex:1]: This auto-generated file is massive. 😮 
But at least it has good docs. 👍

File: lib/app_web/components/core_components.ex [hunk 3]
```diff
@@ -0,0 +1,621 @@
...
+      <.modal id="confirm-modal">
+        Are you sure?
+        <:confirm>OK</:confirm>
+        <:cancel>Cancel</:cancel>
+      </.modal>
+
+  JS commands may be passed to the `:on_cancel` and `on_confirm` attributes
+  for the caller to react to each button press, for example:
+
+      <.modal id="confirm" on_confirm={JS.push("delete")} on_cancel={JS.navigate(~p"/posts")}>
```
Comment (nelsonic) [inline lib/app_web/components/core_components.ex:31]: **`Modals`** are _horrible_ in _most_ cases, don't use them! https://github.com/dwyl/product-ux-research/issues/38

File: lib/app_web/components/core_components.ex [hunk 4]
```diff
@@ -0,0 +1,621 @@
...
+      to: selector,
+      time: 200,
+      transition:
+        {"transition-all transform ease-in duration-200",
+         "opacity-100 translate-y-0 sm:scale-100",
+         "opacity-0 translate-y-4 sm:translate-y-0 sm:scale-95"}
+    )
+  end
+
+  def show_modal(js \\ %JS{}, id) when is_binary(id) do
```
Comment (nelsonic) [inline lib/app_web/components/core_components.ex:561]: don't.

File: priv/gettext/en/LC_MESSAGES/errors.po [hunk 5]
```diff
@@ -50,13 +50,23 @@ msgid "are still associated with this entry"
 msgstr ""
 
 ## From Ecto.Changeset.validate_length/3
+msgid "should have %{count} item(s)"
```
Comment (nelsonic) [inline priv/gettext/en/LC_MESSAGES/errors.po:53]: are we actually _using_ `gettext` for translating the UI? 💭 
If so, we could add a whole new section to this tutorial! 💡

Comment (nelsonic) [inline priv/gettext/en/LC_MESSAGES/errors.po:53]: are we actually _using_ `gettext` for translating the UI? 💭 
If so, we could add a whole new section to this tutorial! 💡

**Review by nelsonic — APPROVED** [review]
@LuchoTurtle this is a great update. Nice work! 🎉

**Review by nelsonic — APPROVED** [review]
@LuchoTurtle this is a great update. Nice work! 🎉

File: lib/app/mailer.ex (hunk 1 above)
Comment (LuchoTurtle) [reply lib/app/mailer.ex:1]: Do you want me to delete these files? Why did you merge if you gave this feedback?

**Pull request merged**